File: Sortings/merge sort.cpp
//
// Created by Codilis on 20/09/2018.
//

#include <iostream>

using namespace std;

void merge(int a[], int l, int m, int r) {


    int s1 = m-l+1;
    int s2 = r-m;

    int L[s1], R[s2];

    for (int n = 0; n < s1; ++n) {
        L[n] = a[l+n];
    }
    for (int n = 0; n < s2; ++n) {
        R[n] = a[m+1+n];
    }

    int i=0, k=l,j=0;
    while (i<s1 && j<s2){
        if(L[i] < R[j]){
            a[k] = L[i];
            i += 1;
        }
        else{
            a[k] = R[j];
            j += 1;
        }
        k++;
    }

    while (i < s1){
        a[k] = L[i];
        i += 1;
        k += 1;
    }

    while (j<s2){
        a[k] = R[j];
        k += 1;
        j += 1;
    }

}

void merge_sort(int a[], int start, int end) {

    if(start < end) {
        int mid = start +(end-start) / 2;
        merge_sort(a, start, mid);
        merge_sort(a, mid+1, end);
        merge(a, start, mid, end);
    }



}

int main(){

    int a[] = {64, 34, 25, 12, 22, 11, 90};
    int end = sizeof(a)/ sizeof(a[0]);
    int start = 0;
    merge_sort(a, start, end-1);
    for (int i = 0; i < end; ++i) {
        cout<<a[i]<<"\t";
    }
    cout<<endl;
    return 0;
}

File: Sortings/merge_sort.py

def merge_sort(a, start, end):
    if end-start == 1:
        return [a[start]]
    if end-start == 2:
        return [min(a[start], a[end-1]), max(a[start], a[end-1])]

    mid = (start+end)//2

    left_arr = merge_sort(a, start, mid+1)
    right_arr = merge_sort(a, mid+1, end)

    left_len = mid+1-start
    right_len = end-mid-1

    k = 0
    i = 0
    j = 0
    while i<left_len and j<right_len:
        if left_arr[i] < right_arr[j]:
            a[k] = left_arr[i]
            k += 1
            i += 1
        elif left_arr[i] > right_arr[j]:
            a[k] = right_arr[j]
            k += 1
            j += 1
        else:
            a[k] = right_arr[j]
            k += 1
            a[k] = left_arr[i]
            k += 1
            i += 1
            j += 1

    while i<left_len:
        a[k] = left_arr[i]
        k += 1
        i += 1

    while j<right_len:
        a[k] = right_arr[j]
        j += 1
        k += 1

    return a




if __name__ == '__main__':
    arrays = [[1, 4, 6, 3, 5, 8, 2], [4, 6, 10, 12, 15, 23, 28, 35], [12, 15, 23, 4 , 6, 10, 35], [35, 28, 23, 15, 12, 10, 6, 4], [12], [12, 4], [12, 15, 23, 4 , 6, 10, 35, 28, 100, 130, 500, 1000, 235, 554, 75, 345, 800, 222, 333, 888, 444, 111, 666, 777, 60], [12, 15, -23, -4 , 6, 10, -35, 28], [12, 12, 23, 4 , 6, 6, 10, -35, 28], [12, 12, 12, 12, 12]]

    for arr in arrays:
        answer = merge_sort(arr, 0, len(arr))
        # print(answer)
